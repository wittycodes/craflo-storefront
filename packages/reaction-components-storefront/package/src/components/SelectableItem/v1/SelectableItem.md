### Overview

A `SelectableItem` is used in a [SelectableList](./#!/SelectableList) and consists of a radio button, label, and optional detail view. The label and detail view can both support text or render other elements, like icons and links.

`SelectableItem` is a fully controlled, stateless component, which means that it will only display as checked when the `isChecked` prop is `true`, even if you click it. The parent of `SelectableItem` is expected to listen for `onChange` and update the `isChecked` prop as appropriate.

An individual item can be checked or unchecked, and read-only or editable:

```jsx noeditor
<SelectableItem label="Default unchecked" value="default" />
```

```jsx noeditor
<SelectableItem isChecked label="Default checked" value="default2" />
```

```jsx noeditor
<SelectableItem isReadOnly label="Default disabled and unchecked" value="default3" />
```

```jsx noeditor
<SelectableItem isChecked isReadOnly label="Default disabled and checked" value="default4" />
```

### Usage
- All labels should be written in sentence caps.
- Radio buttons should be used when only one option can be selected from a set of defined values.

### Types

#### SelectableItem without `detail`

```jsx
<SelectableItem label="Default address" value="default5" />
```

#### SelectableItem with `detail`

Pass any element - text, SVGs, other components, like `Address`,  or any React elements - into `detail` to display a secondary information and action on the right-hand side.

##### Plain text as detail

```jsx
<SelectableItem label="Free shipping" detail="$0.00" value="free"/>
```

```jsx
<SelectableItem label="Free shipping" detail="\u2714" value="free"/>
```

##### Element as detail

```jsx
import Button from "../../Button/v1/Button";
const link = (
  <Button title="Default" className="myBtn" isTextOnly isShortHeight>Default Text</Button>
);

<SelectableItem label="Default" detail={link} value="default8" />
```

##### Address as detail

```jsx
import Address from "../../Address/v1/Address";
const address = {
  _id: "1",
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
};

const addressElement = (
  <Address address={address} />
);

<SelectableItem isStacked label="Entered address:" detail={addressElement} value="default12" />
```

##### Icon as detail
```jsx
const iconClear = require("../../../../../package/src/svg/iconClear").default;

<SelectableItem label="Free shipping" detail={iconClear} value="default9" />
```

##### Secondary icon
```jsx
const iconVisa = require("../../../../../package/src/svg/iconVisa").default;

<SelectableItem isLeftAligned label="Visa" detail="ends in 4040" icon={iconVisa} value="default123" />
```

### Specs

| Property                       | Style                                     |
| ------------------------------ | :---------------------------------------: |
| Label font                     | `@rui-label-text` capitalized             |
| Label font weight              | `400`                                     |
| Label font weight              | `700` when `isStacked` or `isLeftAligned` |
| Label color                    | `@cool-grey-500`                          |
| Border color                   | `@cool-grey-500`                          |
| Radio button color             | `@cool-grey-500`                          |
| Radio button height and width  | 20px                                      |
| Radio button and label spacing | 11px                                      |
| Radio button border            | 2px solid `@cool-grey-500`                |
| Selected circle size           | 11px                                      |
| Cursor                         | Pointer (radio button and label)          |

### Theme

Assume that any theme prop that does not begin with "rui" is within `rui_components`. See [Theming Components](./#!/Theming%20Components).

| Theme Prop                                       | Default                    | Description                                                |
| ------------------------------------------------ | -------------------------- | ---------------------------------------------------------- |
| `SelectableItemRadioButton.backgroundColor`      | white                      | Radio button background color                              |
| `SelectableItemRadioButton.borderColor`          | coolGrey500                | Radio button border color                                  |
| `SelectableItemRadioButton.borderWidth`          | 2px                        | Radio button border width                                  |
| `SelectableItemRadioButton.checkSize`            | 10px                       | Radio button check height and width                        |
| `SelectableItemRadioButton.color`                | coolGrey500                | Radio button foreground color                              |
| `SelectableItemRadioButton.disabledCursor`       | not-allowed                | Cursor when hovering a disabled radio button               |
| `SelectableItemRadioButton.disabledFillColor`    | black10                    | Radio button fill color when disabled                      |
| `SelectableItemRadioButton.disabledOpacity`      | 0.5                        | Radio button opacity when disabled                         |
| `SelectableItemRadioButton.focus`                | `0 0 0 2px ${colors.teal}` | Radio button box shadow when focused                       |
| `SelectableItemRadioButton.focusOutline`         | `1px solid transparent`    | Radio button outline when focused                          |
| `SelectableItemRadioButton.size`                 | 20px                       | Radio button height and width                              |
| `SelectableItemRadioButton.spacingToLabel`       | 10px                       | Space between the radio button and its label               |
| `SelectableList.borderColor`                     | black10                    | Border color when `isBordered`                             |
| `SelectableList.borderRadius`                    | 2px                        | Border radius for all outside corners when `isBordered`    |
| `SelectableList.borderStyle`                     | solid                      | Border style when `isBordered`                             |
| `SelectableList.borderWidth`                     | 1px                        | Border width when `isBordered`                             |
| `SelectableList.height`                          | 50px                       | Height of each item                                        |
| `SelectableList.heightMobile`                    | 60px                       | Height of each item on a small screen                      |
| `SelectableList.iconHeight`                      | 24px                       | Height of the icon                                         |
| `SelectableList.iconLeft`                        | 30px                       | Left offset of the icon                                    |
| `SelectableList.iconSpacingToLabel`              | 10px                       | Space after the icon and before the label                  |
| `SelectableList.iconWidth`                       | 38px                       | Icon width                                                 |
| `SelectableList.itemPaddingBottom`               | 0                          | Bottom padding for each item                               |
| `SelectableList.itemPaddingLeft`                 | 10px                       | Left padding for each item                                 |
| `SelectableList.itemPaddingRight`                | 10px                       | Right padding for each item                                |
| `SelectableList.itemPaddingTop`                  | 0                          | Top padding for each item                                  |
| `SelectableList.leftAlignedLabelFontWeight`      | 700                        | Font weight for the label when `isLeftAligned`             |
| `SelectableList.leftAlignedDetailSpacingToLabel` | 2px                        | Spacing between label and detail text when `isLeftAligned` |
| `SelectableList.stackedSpacingToLabel`           | 30px                       | Spacing between radio button and detail when `isStacked`   |
| `SelectableList.stackedSpacingBelowLabel`        | 10px                       | Spacing between label and detail when `isStacked`          |
#### Typography

- The label text uses `labelText` style with `rui_components.SelectableItemLabel` override
- The detail text when not `isLeftAligned` uses `bodyText` style with `rui_components.SelectableItemDetail` override
- The detail text when `isLeftAligned` uses `labelText` style with `rui_components.SelectableItemDetailLeft` override
