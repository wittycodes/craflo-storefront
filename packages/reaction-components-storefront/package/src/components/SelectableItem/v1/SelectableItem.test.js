import React from "react";
import renderer from "react-test-renderer";
import SelectableItem from "./SelectableItem";

test("basic snapshot with empty props", () => {
  const component = renderer.create(<SelectableItem label="Label" detail="Detail" value="detail" />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
