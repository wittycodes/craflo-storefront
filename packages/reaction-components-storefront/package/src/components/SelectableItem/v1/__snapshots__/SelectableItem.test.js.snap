// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with empty props 1`] = `
.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-font-smoothing: antialiased;
  color: #505558;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 14px;
  font-style: normal;
  font-stretch: normal;
  font-weight: 400;
  -webkit-letter-spacing: .02em;
  -moz-letter-spacing: .02em;
  -ms-letter-spacing: .02em;
  letter-spacing: .02em;
  line-height: 1.25;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ffffff;
  border-radius: 50%;
  border-color: #3c3c3c;
  border-style: solid;
  border-width: 2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 10px;
  margin-top: 0;
  width: 20px;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  cursor: pointer;
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  visibility: visible;
  white-space: nowrap;
  width: 1px;
}

.c1:checked + .c2 .c4::before {
  content: " ";
  display: inline-block;
  position: relative;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background-color: #3c3c3c;
}

.c1:focus + .c2 .c4 {
  box-shadow: 0 0 0 2px #8ce0c9;
  outline: 1px solid transparent;
}

.c1:disabled + .c2 .c4 {
  background-color: #e6e6e6;
}

.c1:disabled + .c2 {
  opacity: .5;
}

.c1:disabled + .c2:hover {
  cursor: not-allowed;
}

.c6 {
  -webkit-font-smoothing: antialiased;
  color: #3c3c3c;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 16px;
  font-style: normal;
  font-stretch: normal;
  font-weight: 400;
  -webkit-letter-spacing: .03em;
  -moz-letter-spacing: .03em;
  -ms-letter-spacing: .03em;
  letter-spacing: .03em;
  line-height: 1.5;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: auto;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: 0;
  margin-top: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 50px;
}

@media (max-width:768px) {
  .c0 {
    height: 60px;
  }
}

<div>
  <div
    className="c0"
  >
    <input
      checked={false}
      className="c1"
      disabled={false}
      id="radio_SelectableItem_1"
      onChange={[Function]}
      type="radio"
      value="detail"
    />
    <label
      className="c2 c3"
      htmlFor="radio_SelectableItem_1"
    >
      <span
        className="c4 c5"
      />
      Label
    </label>
    <div
      className="c6"
    >
      Detail
    </div>
  </div>
</div>
`;
